[MRG + 2] FIX bug in svm's decision values when `decision_function_shape` is `'ovr'`

btdai opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/master/CONTRIBUTING.md#Contributing-Pull-Requests
-->
#### Reference Issue

<!-- Example: Fixes #1234 -->

Fixes #6416
#### What does this implement/fix? Explain your changes.

Changed the sign of dec for the function ovr_decision_function
from
`return _ovr_decision_function(dec < 0, dec, len(self.classes))`
to
`return ovr_decision_function(dec < 0, -dec, len(self.classes))`

Reason: for function _ovr_decision_function, the first argument should be consistent with the second argument, so if we take dec < 0, i.e., set True for negative decision value, the confidence that we feed into _ovr_decision_function needs to be inverted as well.

Hence I proposed the above change. Please kindly refer to the examples in issue #6416, thank you very much. 
#### Any other comments?

<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

the sign of dec values should be changed to feed into function _ovr_decision_function

btdai committed:
changed line 554 of svm/base.py

the sign of dec values should be changed to feed into function _ovr_decision_function

btdai commented:
I hope this PR is correct.

lesteve commented:
It would be great to add a test that is failing on master and passing in this branch. I think the place where it belongs is sklearn/svm/tests/test_svm.py.

amueller commented:
+1 on the test!

raghavrv commented:
Why not change voting to `dec > 0` instead?

raghavrv commented:
Actually changing `dec > 0` is the correct thing to do here I think...

For instance, after changing `dec` to `-dec` in master

```python
import numpy as np
from sklearn.svm import SVC
from sklearn.utils.testing import assert_true

base_points = np.array([[1, 1], [2, 2]])

X = np.vstack((
    base_points * [1, 1],    # Q1
    base_points * [-1, 1],   # Q2
    base_points * [1, -1],   # Q3
    base_points * [-1, -1]   # Q4
    ))

y = [0] * 2 + [1] * 2 + [2] * 2 + [3] * 2

svm = SVC(kernel='linear', decision_function_shape='ovr')
svm.fit(X, y)

# One point close to the decision boundary and another far away
base_points = np.array([[5, 5], [10, 10]])

# For all the quadrants (classes)
X = np.vstack((
    base_points * [1, 1],    # Q1
    base_points * [-1, 1],   # Q2
    base_points * [1, -1],   # Q3
    base_points * [-1, -1]   # Q4
    ))

list(zip(svm.predict(X), svm.decision_function(X)))
```

Gives

```python
[(0, array([ 0.25,  2.  ,  1.  ,  2.75])),
 (0, array([ 0.5,  2. ,  1. ,  2.5])),
 (1, array([ 2.  ,  0.25,  2.75,  1.  ])),
 (1, array([ 2. ,  0.5,  2.5,  1. ])),
 (2, array([ 2.  ,  2.75,  0.25,  1.  ])),
 (2, array([ 2. ,  2.5,  0.5,  1. ])),
 (3, array([ 2.75,  2.  ,  1.  ,  0.25])),
 (3, array([ 2.5,  2. ,  1. ,  0.5]))]
```

You can see that for the first prediction, when predicted class is 0, the confidence is higher for class = 3 and not 0.

If you fix it as 

`return _ovr_decision_function(dec > 0, dec, len(self.classes_))[:, ::-1]`

You get

```python
[(0, array([ 3.25,  1.  ,  2.  , -0.25])),
 (0, array([ 3.5,  1. ,  2. , -0.5])),
 (1, array([ 1.  ,  3.25, -0.25,  2.  ])),
 (1, array([ 1. ,  3.5, -0.5,  2. ])),
 (2, array([ 1.  , -0.25,  3.25,  2.  ])),
 (2, array([ 1. , -0.5,  3.5,  2. ])),
 (3, array([-0.25,  1.  ,  2.  ,  3.25])),
 (3, array([-0.5,  1. ,  2. ,  3.5]))]
```

Which is correct...

And the test should be the above snipped with the last line replaced by 

```python
decisions = svm.decision_function(X)

pos_class_decisions = decisions[range(8), svm.predict(X)].reshape([4, 2])
    
# Test if the point closer to the decision boundary has a lower value
# compared to a poing farther away from the boundary
assert_true(np.all(pos_class_decisions[:, 0] < pos_class_decisions[:, 1]))

# Assert that the predicted class has the maximum value
assert_array_equal(np.argmax(decisions, axis=1), svm.predict(X))
```

btdai commented:
@raghavrv thank you very much for your reply, so you agree with me that this is indeed a bug, and we shall change to 
from
`return _ovr_decision_function(dec < 0, dec, len(self.classes))`
to either
1) `return ovr_decision_function(dec < 0, -dec, len(self.classes))`
or
2) `return ovr_decision_function(dec > 0, dec, len(self.classes))`

I think it should be option 1 and you think it should be option 2.

The reason why I proposed change to option 1 is this: (from line 413 onwards, in master/sklearn/utils/multiclass.py)

```
k = 0
    for i in range(n_classes):
        for j in range(i + 1, n_classes):
            sum_of_confidences[:, i] -= confidences[:, k]
            sum_of_confidences[:, j] += confidences[:, k]
            votes[predictions[:, k] == 0, i] += 1
            votes[predictions[:, k] == 1, j] += 1
            k += 1
```
this piece of code aggregate sum of confidence from each of the n_classes * (n_classes -1) / 2 classifiers. For the k^{th} binary classifier, it's between class i and class j, it minus confidence from class i, and add confidence for class j, so the confidence is actually the confidence in class j, right?

However, the dec values is positive for the first class (class i), and negative for the second class (class j), in the ovo case of SVM, so I think we should change the sign of dec, so that class i should receive negative confidence, class j should receive positive confidence, to match the intuition in the _ovr_decision_function defined in master/sklearn/utils/multiclass.py

Now, let's run an example, with plots of original code (first row), option 1 (second row), and option 2 (third row).

```
import numpy as np
from sklearn import datasets, svm, multiclass
import matplotlib.pyplot as plt

iris = datasets.load_iris()
x = iris.data[:,:2]
y = iris.target

ovo = svm.SVC(kernel='linear')
ovr = svm.SVC(kernel='linear', decision_function_shape='ovr')
ovo.fit(x, y)
ovr.fit(x, y)

test_pts = [()] * 6
test_pts[0] = (5.0, 3.1) # red class nearer to decision boundaries
test_pts[1] = (4.5, 3.5) # red class further from decision boundaries
test_pts[2] = (5.5, 2.6) # green class nearer to decision boundaries
test_pts[3] = (5.0, 2.0) # green class further from decision boundaries
test_pts[4] = (6.5, 3.0) # blue class nearer to decision boundaries
test_pts[5] = (7.5, 3.0) # blue class further from decision boundaries
test_pts = np.array(test_pts)

ovo_deci = ovo.decision_function(test_pts)
ovr_deci = ovr.decision_function(test_pts)
# the line above actually calls multiclass._ovr_decision_function(ovo_deci < 0, ovo_deci, 3)
# I personally believe we should change the sign of ovo_deci as the next line
deci_opt_1 = multiclass._ovr_decision_function(ovo_deci < 0, -ovo_deci, 3)
deci_opt_2 = multiclass._ovr_decision_function(ovo_deci > 0, ovo_deci, 3)

color = ['red', 'green', 'blue']
y_color = [color[i] for i in y]

text_pos = [(5.0,4.5),(5.0,4.8),(7.5,2.1),(7.5,1.8),(7.5,4.8),(7.5,4.5)]

plt.figure(figsize = (18, 18))
for i in range(3):
    plt.subplot(3, 3, i + 1)
    plt.scatter(x[:,0], x[:,1], c=y_color, marker='+')
    plt.scatter(test_pts[2*i:2*i+2,0], test_pts[2*i:2*i+2,1], c=color[i], s = 80)
    for j in range(2*i, 2*i+2):
        plt.annotate('%.3f' % ovr_deci[j,i], xy=test_pts[j], xytext=text_pos[j],arrowprops=dict(facecolor='black', shrink=0.05))
    plt.subplot(3, 3, i + 4)
    plt.scatter(x[:,0], x[:,1], c=y_color, marker='+')
    plt.scatter(test_pts[2*i:2*i+2,0], test_pts[2*i:2*i+2,1], c=color[i], s = 80)
    for j in range(2*i, 2*i+2):
        plt.annotate('%.3f' % deci_opt_1[j,i], xy=test_pts[j], xytext=text_pos[j],arrowprops=dict(facecolor='black', shrink=0.05))
    plt.subplot(3, 3, i + 7)
    plt.scatter(x[:,0], x[:,1], c=y_color, marker='+')
    plt.scatter(test_pts[2*i:2*i+2,0], test_pts[2*i:2*i+2,1], c=color[i], s = 80)
    for j in range(2*i, 2*i+2):
        plt.annotate('%.3f' % deci_opt_2[j,i], xy=test_pts[j], xytext=text_pos[j],arrowprops=dict(facecolor='black', shrink=0.05))

plt.show()
```
![opt_1_or_2](https://cloud.githubusercontent.com/assets/17401504/19718666/dc98f06e-9b98-11e6-95f5-affc07f8f5af.JPG)

btdai commented:
@raghavrv Raghav, I tried your code as well, but I got totally different result, I have used your base points, X and y

```
import numpy as np
from sklearn.svm import SVC
from sklearn.utils.testing import assert_true

base_points = np.array([[1, 1], [2, 2]])

X = np.vstack((
    base_points * [1, 1],    # Q1
    base_points * [-1, 1],   # Q2
    base_points * [1, -1],   # Q3
    base_points * [-1, -1]   # Q4
    ))

y = [0] * 2 + [1] * 2 + [2] * 2 + [3] * 2

svm = SVC(kernel='linear', decision_function_shape='ovr')
svm.fit(X, y)

# One point close to the decision boundary and another far away
base_points = np.array([[5, 5], [10, 10]])

# For all the quadrants (classes)
X = np.vstack((
    base_points * [1, 1],    # Q1
    base_points * [-1, 1],   # Q2
    base_points * [1, -1],   # Q3
    base_points * [-1, -1]   # Q4
    ))

list(zip(svm.predict(X), svm.decision_function(X)))

ovo = SVC(kernel='linear', decision_function_shape='ovo')
ovo.fit(X, y)
ovo_deci = ovo.decision_function(X)
from sklearn import multiclass
deci_opt_1 = multiclass._ovr_decision_function(ovo_deci < 0, -ovo_deci, 4)
deci_opt_2 = multiclass._ovr_decision_function(ovo_deci > 0, ovo_deci, 4)

print(deci_opt_1)
print([np.argmax(i) for i in deci_opt_1])
print(deci_opt_2)
print([np.argmax(i) for i in deci_opt_2])
```

The option 1 and 2 gives:

```
>>> print(deci_opt_1)
[[ 3.25  2.    1.   -0.25]
 [ 3.5   2.    1.   -0.5 ]
 [ 2.    3.25 -0.25  1.  ]
 [ 2.    3.5  -0.5   1.  ]
 [ 2.   -0.25  3.25  1.  ]
 [ 2.   -0.5   3.5   1.  ]
 [-0.25  2.    1.    3.25]
 [-0.5   2.    1.    3.5 ]]
>>> print([np.argmax(i) for i in deci_opt_1])
[0, 0, 1, 1, 2, 2, 3, 3]
>>> print(deci_opt_2)
[[-0.25  2.    1.    3.25]
 [-0.5   2.    1.    3.5 ]
 [ 2.   -0.25  3.25  1.  ]
 [ 2.   -0.5   3.5   1.  ]
 [ 2.    3.25 -0.25  1.  ]
 [ 2.    3.5  -0.5   1.  ]
 [ 3.25  2.    1.   -0.25]
 [ 3.5   2.    1.   -0.5 ]]
>>> print([np.argmax(i) for i in deci_opt_2])
[3, 3, 2, 2, 1, 1, 0, 0]
```
when we use np.argmax, it should be consistent with y. So I think option 1 -- changing sign of dec -- is right, what do you think?

raghavrv commented:
@btdai Thanks for the patient response. I think you are right! Option 1 is the correct way to go here. Apologies for the confusion...

The test should be along the lines of 

```python

import numpy as np
from sklearn.svm import SVC
from sklearn.utils.testing import assert_true
from sklearn.utils.testing import assert_array_equal

base_points = np.array([[1, 1], [2, 2], [2, 1], [1, 2]])

X = np.vstack((
    base_points * [1, 1],    # Q1
    base_points * [-1, 1],   # Q2
    base_points * [-1, -1],  # Q3
    base_points * [1, -1]    # Q4
    ))

y = [1] * 4 + [2] * 4 + [3] * 4 + [4] * 4

# One point close to the decision boundary and another far away
base_points = np.array([[5, 5], [10, 10]])

# For all the quadrants (classes)
X_test = np.vstack((
    base_points * [1, 1],    # Q1
    base_points * [-1, 1],   # Q2
    base_points * [-1, -1],  # Q3
    base_points * [1, -1],   # Q4
    ))

svm = SVC(kernel='linear', decision_function_shape='ovr')
svm.fit(X, y)

decisions = svm.decision_function(X_test)

# Subtract 1 from predictions to get indices for that class (Quadrant)
pos_class_decisions = decisions[range(8), svm.predict(X_test) - 1].reshape([4, 2])

# Test if the point closer to the decision boundary has a lower value
# compared to a poing farther away from the boundary
assert_true(np.all(pos_class_decisions[:, 0] < pos_class_decisions[:, 1]))

# Assert that the predicted class has the maximum value
# Add 1 to convert argmax positions to class labels (Quadrants)
assert_array_equal(np.argmax(decisions, axis=1) + 1, svm.predict(X_test))

# [1000, 0.1] is closer to Q4 compared to Q2/Q3
confidences = svm.decision_function([[1000, 0.1]])[0]

confidence_Q4 = confidences[4 - 1]
confidence_Q3 = confidences[3 - 1]
confidence_Q2 = confidences[2 - 1]

assert_true(confidence_Q4 > confidence_Q3)
assert_true(confidence_Q4 > confidence_Q2)
assert_true(confidence_Q2 > confidence_Q3)
```

If you have a better test, please feel free to use that! :)

This test fails in master as well as with my (incorrect) fix...

amueller commented:
I don't understand how this test tests the right thing. We want a tie between the hard predicted classes on the test point, right? So you should compare [.1, .1] and [1, 1] in all the quadrants and see that the decision function for that quadrant is higher further away from 0,0.
Also, I feel like you overcomplicate things by having four points per class. why not one?

raghavrv commented:
>  So you should compare [.1, .1] and [1, 1] in all the quadrants and see that the decision function for that quadrant is higher further away from 0,0.

It instead checks 
1. If the decision value for `[5, 5]` is less than that for `[10, 10]` in all quadrants. This is not the case currently in master but fixed in this PR... (For 5 decision value is 3 - 0.25 and for 10, decision value is 3 - 0.5)
2. If the decision value for `[1000, 0.1]` (which causes a tie in some classes but not all), indicates that it lies in Quadrant 1, is closer to Q4, than to Q2/Q3. This also is not the case in master and is fixed by this PR.

> I feel like you over complicate things by having four points per class. why not one?

True we need not have 4 points to train. Just the [-1, 1] etc would suffice...

btdai commented:
Thank you @raghavrv , I think your test case is good, but I'm still new to scikit-learn community, I'm not sure if @amueller Andreas' idea is to have one test case added into test case. I think your test case is good, but I made X and X_test of the same length, so we can compare the array predictions with y directly.

This piece of code can't run on my scikit-learn (mine is not developer version), could you help me run see if the current code runs on your development mode.

Also, I think we should assert the deci_vals for positive class greater than 0 as well.

```
import numpy as np
from sklearn.svm import SVC
from sklearn.utils.testing import assert_true
from sklearn.utils.testing import assert_array_equal

base_points = np.array([[1, 2], [2, 1]])

X = np.vstack((
    base_points * [1, 1],    # Q1
    base_points * [-1, 1],   # Q2
    base_points * [-1, -1],  # Q3
    base_points * [1, -1]    # Q4
    ))

y = [0] * 2 + [1] * 2 + [2] * 2 + [3] * 2

# One point close to the decision boundary and another far away
base_points = np.array([[5, 5], [10, 10]])

# For all the quadrants (classes)
X_test = np.vstack((
    base_points * [1, 1],    # Q1
    base_points * [-1, 1],   # Q2
    base_points * [-1, -1],  # Q3
    base_points * [1, -1],   # Q4
    ))

svm = SVC(kernel='linear', decision_function_shape='ovr')
svm.fit(X, y)

predictions = svm.predict(X_test)

# Test if the prediction is the same as y
assert_array_equal(predictions, y)

deci_vals = svm.decision_function(X_test)

# Assert that the predicted class has the maximum value
assert_array_equal(np.argmax(deci_vals, axis=1), svm.predict(X_test))

# Get decision value at test points for the predicted class
pos_class_deci_vals = deci_vals[range(8), svm.predict(X_test)].reshape([4, 2])

# we shall also assert pos_class_deci_vals > 0 here

# Test if the point closer to the decision boundary has a lower decision value
# compared to a point farther away from the boundary
assert_true(np.all(pos_class_deci_vals[:, 0] < pos_class_deci_vals[:, 1]))
```

amueller commented:
@raghavrv ok, I didn't entirely understand your test case. I'm good with just including it, commenting it thoroughly and making it as succinct as possible.

amueller commented:
@btdai would you like to finish this up? I'd like to get this into the 0.18.1 release and it's the last blocking issue. Otherwise maybe @raghavrv wants to pick it up?

btdai commented:
@amueller I'm sorry to be the last one blocking this issue. When would you like to release 0.18.1 ?

btdai commented:
sorry to ask again, I see the there's test_multiclass.py in master/sklearn/tests, do I need to create another test just for this function?

jnothman commented:
would like to release ASAP. The test should be in `sklearn/svm/tests/test_svm.py`

btdai commented:
Hi, I have created this PR: https://github.com/scikit-learn/scikit-learn/pull/7810

it added the test file test_ovr_decision_function.py in sklearn/svm/tests

Please review this PR, thank you.

lesteve commented:
You need to add the test in this PR and not create a separate PR, I will close #7810. Also could you add the tests in  `sklearn/svm/tests/test_svm.py` as @jnothman mentioned above and not in a separate file.

For this you just need to create an additional commit in this branch i.e. btdai_patch_ovr_decision_function and then use git push to update the branch on your fork. Sorry if it is not clear enough, maybe try to read the links about github that we gave you before again. Do say if you get completely stuck though.

btdai committed:
Create test_ovr_decision_function.py

btdai commented:
@lesteve thank you for your comments, yah, i should have just added the test case in this branch.

I've now added it, seems it's going through some automatic checks.

lesteve commented:
Please add your `test_ovr_decision_function` function in `sklearn/svm/tests/test_svm.py` and do not create a new file (sklearn/svm/tests/test_ovr_decision_function.py).

Also Travis is failing because of flake8 issues, if you are on Linux or OSX you can run `bash build_tools/travis/flake8_diff.sh` locally and fix the errors. They are mostly unused imports and whitespace issues.

btdai committed:
Delete test_ovr_decision_function.py

btdai committed:
Update test_svm.py

btdai committed:
Update test_svm.py

btdai committed:
Update test_svm.py

btdai committed:
Update test_svm.py

btdai commented:
Thanks @lesteve , I have

1. deleted sklearn/svm/tests/test_ovr_decision_function.py
2. edited sklearn/svm/tests/test_svm.py

flake8 complained about lines longer than 80 chars, corrected too.

btdai commented:
Ok, finally, checks have passed.

@amueller Would you please merge into master?

and thank you @amueller @raghavrv @lesteve @jnothman very much for your help!

lesteve commented:
Thanks for sticking to this PR through the difficulties! I edited your PR title to start with `[MRG]` since it is ready for further review. `[WIP]` indicates that you are still working on it. See https://github.com/scikit-learn/scikit-learn/blob/master/CONTRIBUTING.md#pull-request-checklist for more details.

lesteve commented:
We probably nead an entry in whats_new.rst as well. It's probably quicker if @raghavrv or @amueller do it and push into this PR branch directly.

raghavrv commented on sklearn/svm/tests/test_svm.py:
```diff
@@ -949,3 +949,53 @@ def test_decision_function_shape_two_class():
+    train_base_points = np.array([[1, 2], [2, 1]])
+
+    # For all the quadrants (classes)
+    x = np.vstack((
```
like @amueller says could you have 

```py
X = ([1, 1], [-1, 1], [-1, -1], [1, -1])
y = [0, 1, 2, 3]
```

btdai replied:
@raghavrv yes, @amueller suggested so, but keeping two points (these two are support vectors) per class is to ensure the direction of decision boundaries are 45 degree with the coordinates, so then we can say the decision value at (10, 10) should be greater than that at (5, 5).

raghavrv commented on sklearn/svm/tests/test_svm.py:
```diff
@@ -949,3 +949,53 @@ def test_decision_function_shape_two_class():
+    y_test = [0] * 2 + [1] * 2 + [2] * 2 + [3] * 2
+
+    clf = svm.SVC(kernel='linear', decision_function_shape='ovr')
+    clf.fit(x, y)
```
capital `X`

btdai replied:
changed to X, capital for both X and X_test

raghavrv commented on sklearn/svm/tests/test_svm.py:
```diff
@@ -949,3 +949,53 @@ def test_decision_function_shape_two_class():
+    test_base_points = np.array([[5, 5], [10, 10]])
+
+    # For all the quadrants (classes)
+    x_test = np.vstack((
```
`X_test`

btdai replied:
changed to X_test

raghavrv reviewed (changes requested):
With these nitpicks and a whatsnew entry added, this should be ready for merge I think.

btdai committed:
Update test_svm.py

btdai committed:
Update test_svm.py

btdai commented:
Done with modification, thank you.

lesteve commented:
Still need a whats_new entry, i.e. you need to add a few lines about your fix in doc/whats_new.rst in the 0.18.1 section.

btdai committed:
Update whats_new.rst

btdai commented:
@lesteve I have added 4 lines in doc/whats_new.rst

Thanks

raghavrv commented on doc/whats_new.rst:
```diff
@@ -90,6 +90,10 @@ Bug fixes
 
+   - Fix bug svm's decision values when `decision_function_shape` is
+     `'ovr'` :class:`svm.SVC`.
+     `#7724 <https://github.com/scikit-learn/scikit-learn/pull/7724>`_
```
You could use the new `:issue:` tag for this? Ref: #7657

btdai replied:
@raghavrv sorry, I didn't understand this message, what do you mean by use the new :issue: tag, must i create a new issue?

btdai replied:
Hi, @raghavrv , I used this line below: 
:issue:`7724` by `Bing Tian Dai`.

is it correct?

raghavrv commented on sklearn/svm/tests/test_svm.py:
```diff
@@ -949,3 +949,53 @@ def test_decision_function_shape_two_class():
+        train_base_points * [1, -1]    # Q4
+        ))
+
+    Y = [0, 1, 2, 3]
```
We generally have the convention of 

X for data
y for single target
Y multi target

So it would be nice to use capital `X` and small `y` in this test...

raghavrv commented on sklearn/svm/tests/test_svm.py:
```diff
@@ -949,3 +949,53 @@ def test_decision_function_shape_two_class():
+    train_base_points = np.array([[1, 1]])
+
+    # For all the quadrants (classes)
+    X = np.vstack((
```
I suggest that you can simply replace this by

```py
X = np.array([[1, 1], [-1, 1], [-1, -1], [1, -1]])
```

raghavrv commented on sklearn/svm/tests/test_svm.py:
```diff
@@ -949,3 +949,53 @@ def test_decision_function_shape_two_class():
+        test_base_points * [1, -1]    # Q4
+        ))
+
+    Y_test = [0] * 2 + [1] * 2 + [2] * 2 + [3] * 2
```
small y

raghavrv commented on doc/whats_new.rst:
```diff
@@ -90,6 +90,10 @@ Bug fixes
+   - Fix bug svm's decision values when `decision_function_shape` is
+     `'ovr'` :class:`svm.SVC`.
+     `#7724 <https://github.com/scikit-learn/scikit-learn/pull/7724>`_
+     by `Bing Tian Dai`_
```
This expects that your github / website is present at the end of the file...

raghavrv reviewed (changes requested):
Could you address a few more nitpicks please? :)

jnothman commented on doc/whats_new.rst:
```diff
@@ -90,6 +90,10 @@ Bug fixes
      `#6497 <https://github.com/scikit-learn/scikit-learn/pull/6497>`_
      by `Sebastian Säger`_
 
+   - Fix bug svm's decision values when `decision_function_shape` is
```
Should use double-backticks

jnothman commented on doc/whats_new.rst:
```diff
@@ -90,6 +90,10 @@ Bug fixes
      by `Sebastian Säger`_
 
+   - Fix bug svm's decision values when `decision_function_shape` is
+     `'ovr'` :class:`svm.SVC`.
```
double-backticks, and needs "in"

jnothman replied:
Please add that SVC's `decision_function` was incorrect for versions 0.17.0 through 0.18.0.

jnothman reviewed (changes requested):
minor changes on top of @raghavrv's

btdai committed:
Update test_svm.py

btdai committed:
Update whats_new.rst

jnothman commented on doc/whats_new.rst:
```diff
@@ -90,6 +90,11 @@ Bug fixes
+     is ``ovr`` in :class:`svm.SVC`.
+     :class:`svm.SVC`'s decision_function was incorrect from versions
+     0.17.0 through 0.18.0.
+     :issue:`7724` by `Bing Tian Dai`.
```
The link to your name should be followed by `_` and your name / URL listed at the bottom of the file.

btdai committed:
Update whats_new.rst

btdai commented:
Thanks @jnothman , I have added the underscore

jnothman commented on doc/whats_new.rst:
```diff
@@ -90,6 +90,11 @@ Bug fixes
+     is ``ovr`` in :class:`svm.SVC`.
+     :class:`svm.SVC`'s decision_function was incorrect from versions
+     0.17.0 through 0.18.0.
+     :issue:`7724` by `Bing Tian Dai`_.
```
but you've not added a link at the bottom of the file.

btdai replied:
Got error after I add my name at the bottom of the file :(

How can it be...

btdai committed:
Update whats_new.rst

btdai committed:
Update whats_new.rst

btdai committed:
Update whats_new.rst

raghavrv commented:
Needs a rebase. Then good for merge. Thanks @btdai

btdai committed:
whats_new edited

btdai committed:
Merge branch 'btdai_patch_ovr_decision_function' of https://github.com/btdai/scikit-learn into btdai_patch_ovr_decision_function

jnothman commented:
Sorry, still needs a rebase.

jnothman commented:
i.e. you probably need to update your local copy of master, then merge or rebase with it.

btdai commented:
ya, git rebase btdai_patch_....
 always give me error

btdai committed:
whats_new.rst edited

btdai commented:
bingtian@ASPIRES7:~/git$ git rebase btdai_patch_ovr_decision_function
Current branch btdai_patch_ovr_decision_function is up to date.

rebase says up to date

so I push with
git push https://github.com/btdai/scikit-learn.git btdai_patch_ovr_decision_function

Still can't make it

jnothman commented:
You want something like:
```
git checkout master
gut pull https://github.com/scikit-learn/scikit-learn master
git checkout btdai_patch_ovr_decision_function
git merge master  # or git rebase master, but I think merge should work
# resolve merge conflicts and add conflicted files, then
git commit
git push
```

btdai commented:
Hi, I git rebase with -i

but then I tried to push, it said "Everything up-to-date"

is it normal?

Thanks.

Please help @amueller @raghavrv @jnothman 

bingtian@ASPIRES7:~/git$ git rebase -i btdai_patch_ovr_decision_function
Successfully rebased and updated refs/heads/btdai_patch_ovr_decision_function.
bingtian@ASPIRES7:~/git$ git push https://github.com/btdai/scikit-learn.git btdai_patch_ovr_decision_function
Username for 'https://github.com': btdai
Password for 'https://btdai@github.com':
Everything up-to-date
bingtian@ASPIRES7:~/git$ git push --force https://github.com/btdai/scikit-learn.git btdai_patch_ovr_decision_function
Username for 'https://github.com': btdai
Password for 'https://btdai@github.com':
Everything up-to-date

btdai committed:
Merge branch 'master' of https://github.com/scikit-learn/scikit-learn into btdai_patch_ovr_decision_function

Conflicts:
	doc/whats_new.rst

btdai commented:
Ok, I should have done

git merge master

Thank you @jnothman

jnothman commented:
`git rebase master` would have worked fine, together with a force-push. We are used to asking for rebases, because it was more necessary before some (relatively) recent changes to github. I think merges are fine now.

jnothman commented on doc/whats_new.rst:
```diff
@@ -4881,8 +4886,12 @@ David Huard, Dave Morrill, Ed Schofield, Travis Oliphant, Pearu Peterson.
 
 .. _Peng Meng: https://github.com/mpjlu
 
+<<<<<<< HEAD
```
You've not correctly resolved merge conflicts

btdai committed:
whats_new.rst updated

btdai committed:
Update whats_new.rst

jnothman commented:
Thanks, @btdai!

btdai commented:
thank you for your help, @jnothman

I can work on the next bug now ：)
